<div style="border:solid Chocolate 2px; padding: 40px">

<b>Имран, приветствую!</b>

Буду рада с тобой познакомиться. Меня зовут Евгения и я буду ревьюером твоего проекта. Если ты не против, то предлагаю общаться на "ты":) Если удобнее на "вы", то нет проблем, только скажи об этом.

В процессе работы ты встретишь комментарии <font color='green'>зеленого</font>, <font color='gold'>желтого</font> и <font color='red'>красного</font> цветов. Что они значат, смотри ниже:
    <br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Решение на этом шаге станет еще лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

    
<b>Необходимо исправить❌:</b>
 Решение на этом шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>
    
Увидев мой комментарий, не удаляй его, он будет очень полезен в случае повторной проверки работы😊
    
Свой комментарий помечай так, как показано ниже, тогда нам будет легче вести диалог, а мне выполнить повторную проверку проекта.
    <div class="alert alert-info"> <b>Комментарий студента: </b> </div>
    


# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Имран, молодец, отлично выполнил предобработку!
</div>

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_pivot = data.pivot_table(index = 'children', columns = 'debt', values = 'gender', aggfunc = 'count')
data_pivot['debt_percent'] = data_pivot[1]/(data_pivot[0] + data_pivot[1])
display(data_pivot)

debt            0       1  debt_percent
children                               
0         13028.0  1063.0      0.075438
1          4364.0   444.0      0.092346
2          1858.0   194.0      0.094542
3           303.0    27.0      0.081818
4            37.0     4.0      0.097561
5             9.0     NaN           NaN

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Cупер, верно сгруппировал данные для анализа!

</div>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Рекомендую использовать функцию `display()` вместо `print()` для вывода таблиц в Jupyter Notebook.
</div>

 <div class="alert alert-info"> <b>Комментарий студента: </b> 
    
Спасибо за комментарий про dislpay - запомнил</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Вот, так лучше!

</div>

**Вывод: прослеживается зависимость между наличием детей и уровнем возврата кредита в срок - у заемщиков, которые не имеют детей уровень просрочки составляет 7,5%, тогда, как заемщики с 1 или 2-мя детьми (самые массовые категории с детьми) имеют уровень просрочки более 9%.
При этом, можно отметить, что заемщики с 3-5 детьми выбиваются из общей картины, но их доли не существенны, и, следовательно не репрезентативны для общих выводов.** 

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Ты правильно обратил внимание на то, что выборки несбалансированны и сложно сделать общий вывод. Поэтому лучше, для выявления зависимости, сравнить самые многочисленные категории - семьи с 0, 1 и 2 детьми. У тебя не хватает общего вывода, если мы все обобщим то, что наблюдаем с ростом количества детей в семьях?
</div>

 <div class="alert alert-info"> <b>Комментарий студента: </b> 
        
Дополненный вывод - если сравнить самые многочисленные категории заемщиков - без детей, с 1 ребенком и 2-мя детьми, выходит следующее:
    
1) наблюдается значительная разница в уровне просроченных платежей между заемщиками без детей и с 1 ребенком: 7,5% против 9,2%. Выходит, что заемщик с 1 ребенком относительно на 26% хуже соблюдает свои обязательства, чем заемщики без детей.

2) если сравнить заемщиков с 1 ребенком и 2-мя детьми: 9,23% против 9,45% - можно сделать вывод, что заемщики с двумя детьми немного хуже выполняют свои обязательства, на первый взгляд может показаться, что с ростом кол-ва детей ухудшается платежная дисциплина заемщиков, однако для однозначного вывода недостаточно данных по заемщикам с 3-мя, 4-мя детьми и тд  </div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Отлично!

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
data_pivot_fam = data.pivot_table(index = 'family_status', columns = 'debt', values = 'gender', aggfunc = 'count')
data_pivot_fam['debt_percent'] = data_pivot_fam[1]/(data_pivot_fam[0] + data_pivot_fam[1])
display(data_pivot_fam)

debt                       0    1  debt_percent
family_status                                  
Не женат / не замужем   2523  273      0.097639
в разводе               1105   84      0.070648
вдовец / вдова           888   63      0.066246
гражданский брак        3749  385      0.093130
женат / замужем        11334  927      0.075606

**Вывод: в данном случае можно подвести следующий итог:
1) женатые, а также разведенные - имеет более низкий уровень просрочки;

2) не женатые, а также проживающие в гражданском браке - значительно хуже дисциплинированны в части соблюдения своих обязательств;

3) самый лучший показатель соблюдения сроков платежа - у вдовствующих заемщиков, но они являются самой малочисленной категорией. В целом напрашивается вывод, что заемщики имеющие/имевшие узаконенные семейные отношения являются более дисциплинированными в части соблюдения своих обязательств** 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Абсолютно правильно, заемщики состоящие или состоявшие в браке реже допускают просрочку!

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [29]:
data_pivot_inc = data.pivot_table(index = 'total_income_category', columns = 'debt', values = 'gender', aggfunc = 'count')
data_pivot_inc['debt_percent'] = data_pivot_inc[1]/(data_pivot_inc[0] + data_pivot_inc[1])
display(data_pivot_inc)


# для интереса разобью категорию С на 3 (С1, С2, C3) и B аналогично:
def categorize_income_c(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 100000:
            return 'C3'
        elif 100001 <= income <= 150000:
            return 'C2'
        elif 150001 <= income <= 200000:
            return 'C1'
        elif 200001 <= income <= 300000:
            return 'B3'
        elif 300001 <= income <= 500000:
            return 'B2'
        elif 500001 <= income <= 1000000:
            return 'B1'
        elif income >= 1000001:
            return 'A'
    except:
        pass
data['total_income_category_c'] = data['total_income'].apply(categorize_income_c)

data_pivot_inc_c = data.pivot_table(index = 'total_income_category_c', columns = 'debt', values = 'gender', aggfunc = 'count')
data_pivot_inc_c['debt_percent'] = data_pivot_inc_c[1]/(data_pivot_inc_c[0] + data_pivot_inc_c[1])
display(data_pivot_inc_c)

debt                       0     1  debt_percent
total_income_category                           
A                         23     2      0.080000
B                       4660   354      0.070602
C                      14568  1353      0.084982
D                        328    21      0.060172
E                         20     2      0.090909

debt                        0    1  debt_percent
total_income_category_c                         
A                          23    2      0.080000
B1                        185   12      0.060914
B2                       1160   92      0.073482
B3                       3315  250      0.070126
C1                       4335  403      0.085057
C2                       6491  619      0.087060
C3                       3742  331      0.081267
D                         328   21      0.060172
E                          20    2      0.090909

**Выводы: зависимость есть

1) категория С получилось слишком объемной - от 50 тыс до 200 тыс - здесь явно необходимо ее декомпозировать;

2) если работать с тем, что есть выходит, что категория D (при небольшом доходе) имеет более низкий уровень просрочки, однако она недостаточно объёмна, чтобы можно было сделать однозначные выводы 

При более детальной разбивке, среди значимых по объему сегментов:
1) самая дисциплинированная категория - B3 (200-300 тыс)

2) самый высокий риск (а также и доля заемщиков) - в категории С2 (100-150 тыс) ** 

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Какую зависимость мы наблюдаем, если сравним самые многочисленные категории?
</div>

 <div class="alert alert-info"> <b>Комментарий студента: </b> 
        
Дополненный вывод - если сравнить самые многочисленные категории заемщиков - c доходом от 200 000 до 1000 000 (B) и от 50 000 до 200 000 (С):
    
1) Категория B является наиболее дисциплинированной в погашения платежей

2) Категория С является самой многочисленной, при этом имея практически самый худший уровень соблюдения своих обязательств. Учитывая, что данная категория занимает около 75% от всех клиентов, здесь необходим дополнительный анализ в связке с другими параметрами  </div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Принято!

</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_pivot_p = data.pivot_table(index = 'purpose_category', columns = 'debt', values = 'gender', aggfunc = 'count')
data_pivot_p['debt_percent'] = data_pivot_p[1]/(data_pivot_p[0] + data_pivot_p[1])
print(data_pivot_p)

debt                         0    1  debt_percent
purpose_category                                 
операции с автомобилем    3879  400      0.093480
операции с недвижимостью  9971  780      0.072551
получение образования     3619  369      0.092528
проведение свадьбы        2130  183      0.079118


**Вывод: зависимость есть
Заёмщики с целями автомобиль и образование хуже соблюдают свои обязательства, гипотетически по следующим причинам:
1) авто - появляются расходы на содержание авто, а также вероятность форс-мажоров и непредвиденных трат

2) образование - возможны проблемы с монетизацией полученного образования в начале деятельности
** 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец, верно определил более и менее рискованные для банка категории заемщиков по целям кредита!
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: есть пара наиболее вероятных вариантов:
1) либо заемщик никогда не работал, и поэтому нет данных о стаже и доходе;

2) либо заемщик на данный момент безработный, а данные о ранней деятельности не попали в период запроса датасета
* 

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Ты правильно указала человеческий фактор, как наиболее возможную причину образования пропусков в данных. Но какая причина еще может вызвать потерю данных? Например, если произошел какой-то сбой в системе.
</div>

 <div class="alert alert-info"> <b>Комментарий студента: </b> 
        
Да, совсем не подумал о техническом аспекте проблемы с пропущенными значениями  </div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Обычно, это два фактора - человеческий и технический!

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ: дело в том, что медиана дает более точную картину, для понимания общей ситуации, нежели среднее значение. Среднее значение может дать искаженную картину, когда несколько аномальных значений может сиильно повлиять на общую картину, тогда, как с медианой такого влияния не будет.* 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Правильно, медиана менее чувствительна к выбросам!
</div>

### Шаг 4: общий вывод.

Итого по изученным параметрам:
1) Наличие детей - заемщики без детей имеют более высокий риск просрочки платежа;

2) Семейное положение - заемщики, официально не состоящие и не состоявшие в браке - имеют более высокий риск просрочки платежа;

3) Уровень дохода - здесь картина получается не такой однозначной, но все же, сегмент С (а особенно С1 и С2) является более рисковым с точки зрения просрочки платежей;

4) Цели кредита - Авто и Образование несут существенно более риски просрочки платежей.

Анализ необходимо углублять, т.к. наверняка больше инсайдов появится при кросс-анализе параметров. 

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
<b>Необходимо исправить❌:</b>
Имран, общий вывод - обязательный и важный элемент работы. Его нужно сформулировать, взяв основные результаты из всех шагов проекта. Начать нужно с того, с какими данными мы работали и какие действия провели на этапе предобработки, а закончить выявленными при ответе на вопросы зависимостями и рекомендациями заказчику по улучшению сбора данных и скоринга. Заказчик, зачастую, читает только общий вывод исследования - поэтому он должен быть максимально информативным. 
</div>

<div class="alert alert-info"> <b>Комментарий студента: </b> 
        
Общий вывод, попытка 2

Задача: определить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.
    
Данные для анализа: Таблица на 21525 строк и 12 столбцов:
    
    -children (количество детей)
    -days_employed (кол-во дней рабочего стажа)
    -dob_years (возраст клиента в годах)
    -education (уровень образования клиента)
    -education_id (идентификатор уровня образования клиента)
    -family_status (семейное положение)
    -family_status_id (идентификатор семейного положения)
    -gender (пол клиента)
    -income_type (тип занятости)
    -debt (признак наличия просроченной задолженности в истории платежей)
    -total_income (ежемесячный доход)
    -purpose (цель кредита)
        
На этапе предобработки были выполнены следующие действия с данными:
    
    1. Обнаружены пустые значения для столбцов days_employed и total_income в кол-ве 2174. Для столбца days_employed также обнаружены аномальные значения (отрицательные):
    а) пустые значения в столбце total_income были заполнены медианным значением соответствущего income_type;
    б) отрицательные значения в столбце days_employed были заменены на положительные;
    в) пустые значения в столбце days_employed были заполнены медианным значениям соответствущего income_type.
    
    2. Выявлены и удалены анмальные значения столбца children (отрицательное и подозрительно большое - 20).
    
    3. Изменен формат данных столбца total_income - с вещественного (float) на целочисленный (int).
    
    4. На основе столбца purpose создан новый столбце - purpose_category, в котором различные цели кредита были объеденены в несколько основных групп.
    
    5. На основе столбца total_income были созданы столбцы total_income_category и total_income_category_c, в которых различные уровни доходов клиента были объеденены в несколько основых групп.
    
    
По итогам анализа сформированы следующие выводы:
    
    1) Наличие детей
    -наблюдается значительная разница в уровне просроченных платежей между заемщиками без детей и с 1 ребенком: 7,5% против 9,2%. Выходит, что заемщик с 1 ребенком относительно на 26% хуже соблюдает свои обязательства, чем заемщики без детей.
    -если сравнить заемщиков с 1 ребенком и 2-мя детьми: 9,23% против 9,45% - можно сделать вывод, что заемщики с двумя детьми немного хуже выполняют свои обязательства, на первый взгляд может показаться, что с ростом кол-ва детей ухудшается платежная дисциплина заемщиков, однако для однозначного вывода недостаточно данных по заемщикам с 3-мя, 4-мя детьми и тд

    2) Семейное положение 
    -женатые, а также разведенные - имеет более низкий уровень просрочки, 7% и 7.6% соответственно;
    -не женатые, а также проживающие в гражданском браке - значительно хуже дисциплинированны в части соблюдения своих обязательств, 9.8% и 9.3% соответственно;
    -самый лучший показатель соблюдения сроков платежа - у вдовствующих заемщиков (6.6%), но они являются самой малочисленной категорией. В целом напрашивается вывод, что заемщики имеющие/имевшие узаконенные семейные отношения являются более дисциплинированными в части соблюдения своих обязательств.
    
    3) Уровень дохода, если сравнить самые многочисленные категории заемщиков - c доходом от 200 000 до 1000 000 (B) и от 50 000 до 200 000 (С)
    -Категория B является наиболее дисциплинированной в погашения платежей - уровень просрочки 7.1% 
    -Категория С является самой многочисленной, при этом имея практически самый худший уровень соблюдения своих обязательств - уровень просрочки составлет 8.5%. Учитывая, что данная категория занимает около 75% от всех клиентов, здесь необходим дополнительный анализ в связке с другими параметрами.
    
    4) Цель кредита
    - из 4х основных категорий по целям кредита, наиболее высокие риски просрочки у категорий 'операции с авто' и 'получение образования' - около 9.3% просрочки платежей против примерно 7.5% просрочки платежей у остальных категорий. 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Имран, привет! Молодец, оперативно и верно доработал проект. Здорово, что учел не только замечания, но и рекомендации. Благодарю за обратную связь, работа принята, удачи в обучении!)

</div>

<div style="border:solid Chocolate 2px; padding: 40px">
Имран, давай подведем итоги. Ты большой молодец, что старался и проделал такую большую работу. Шаг 1 и 2 ты выполнил без проблем, что очень хорошо:)
    
    Разберем над чем стоит поработать:
    
    1. Оформить общий вывод.
    
    2. Рекомендую:
    - учитывать несбалансированность выборок;
    - использовать функцию `display()` для вывода таблиц;
    - указать еще один фактор, влияющий на появление пропусков.
    
    Пересмотри работу и исправь ошибки. Обязательно пиши, если я могу тебе еще чем-то помочь. Удачи!)
  